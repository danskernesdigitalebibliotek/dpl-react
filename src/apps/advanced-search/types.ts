import { MultiselectOption } from "../../components/multiselect/types";

export const advancedSearchIndexes = [
  "all",
  "term.creatorcontributor",
  "term.subject",
  "term.genreandform",
  "term.mainlanguage",
  "datefirstedition",
  "term.creator",
  "term.title",
  "term.source",
  "dk5",
  "term.specificmaterialtype",
  "term.childrenoradults",
  "term.publisher",
  "term.isbn"
] as const;

export const advancedSearchIndexTranslations = {
  all: "advancedSearchAllIndexesText",
  "term.creatorcontributor": "advancedSearchCreatorText",
  "term.subject": "advancedSearchSubjectText",
  "term.genreandform": "advancedSearchGenreText",
  "term.mainlanguage": "advancedSearchLanguageText",
  datefirstedition: "advancedSearchDateText",
  "term.creator": "advancedSearchMainCreatorText",
  "term.title": "advancedSearchMainTitleText",
  "term.source": "advancedSearchSourceText",
  dk5: "advancedSearchDecimalDk5Text",
  "term.specificmaterialtype": "advancedSearchTypeText",
  "term.childrenoradults": "advancedSearchAudienceText",
  "term.publisher": "advancedSearchPublisherText",
  "term.isbn": "advancedSearchIdentifierText"
} as const;

export type AdvancedSearchIndex = (typeof advancedSearchIndexes)[number];

export type AdvancedSearchClause = {
  value: "AND" | "OR" | "NOT";
  translation: string;
};

export const advancedSearchClauses: AdvancedSearchClause[] = [
  { value: "AND", translation: "clauseAndText" },
  { value: "OR", translation: "clauseOrText" },
  { value: "NOT", translation: "clauseNotText" }
];

export type AdvancedSearchRowData = {
  term: string;
  searchIndex: AdvancedSearchIndex | "";
  clause: AdvancedSearchClause;
  id: number;
};

export type AdvancedSearchRowUpdateRowAspect =
  | "term"
  | "searchIndex"
  | "clause";

export type AdvancedSearchFilterData = {
  materialTypes: MultiselectOption[];
  fiction: MultiselectOption[];
  accessibility: MultiselectOption[];
};

export type AdvancedSearchQuery = {
  rows: AdvancedSearchRowData[];
  filters: AdvancedSearchFilterData;
};

export const initialAdvancedSearchQuery: AdvancedSearchQuery = {
  rows: [
    { term: "", searchIndex: "all", clause: advancedSearchClauses[0], id: 0 },
    { term: "", searchIndex: "all", clause: advancedSearchClauses[0], id: 1 }
  ],
  filters: {
    materialTypes: [{ item: "multiselectAllOptionText", value: "all" }],
    fiction: [{ item: "multiselectAllOptionText", value: "all" }],
    accessibility: [{ item: "multiselectAllOptionText", value: "all" }]
  }
};

export const advancedSearchMaterialTypes: MultiselectOption[] = [
  { item: "advancedSearchFilterBookText", value: "bøger" },
  { item: "advancedSearchFilterEbookText", value: "e-bøger" },
  { item: "advancedSearchFilterAudioBookText", value: "lydbøger" },
  { item: "advancedSearchFilterArticleText", value: "artikler" },
  { item: "advancedSearchFilterMovieText", value: "film" },
  { item: "advancedSearchFilterMusicText", value: "musik" }
];

export const advancedSearchAccessibility: MultiselectOption[] = [
  { item: "advancedSearchFilterPhysicalText", value: "fysisk" },
  { item: "advancedSearchFilterOnlineText", value: "online" }
];

export const advancedSearchFiction: MultiselectOption[] = [
  { item: "advancedSearchFilterFictionText", value: "fiction" },
  { item: "advancedSearchFilterNonFictionText", value: "nonfiction" }
];

export const advancedSearchFilters = {
  materialTypes: "term.generalmaterialtype",
  fiction: "term.fictionnonfiction",
  accessibility: "term.accesstype"
};
