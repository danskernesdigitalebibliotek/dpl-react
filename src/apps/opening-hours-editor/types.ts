export type OpeningHoursCategoriesType = {
  title: string;
  color: string;
};
