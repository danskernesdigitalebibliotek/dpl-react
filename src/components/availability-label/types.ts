// Publizon API useGetV1LoanstatusIdentifier return data a property "loanStatus"
// which can be a number between 0 - 7. These are codes for specific statuses
// a given product can be in - e.g. code 5 means that there is a reservation queue
// on the product.
// #5 is also the only one we know the exact meaning of. The rest of the numbers
// should be updated in the future when we figure out what they mean.
export const publizonProductStatuses = {
  0: {
    isAvailable: true,
    meaning: "Unknown"
  },
  1: {
    isAvailable: true,
    meaning: "Unknown"
  },
  2: {
    isAvailable: true,
    meaning: "Unknown"
  },
  3: {
    isAvailable: true,
    meaning: "Unknown"
  },
  4: {
    isAvailable: true,
    meaning: "Reservable"
  },
  5: {
    isAvailable: false,
    meaning: "Reservation queue on the material"
  },
  6: {
    isAvailable: true,
    meaning: "Unknown"
  },
  7: {
    isAvailable: true,
    meaning: "Unknown"
  }
};

export default {};
