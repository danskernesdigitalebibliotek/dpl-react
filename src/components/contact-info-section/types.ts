export interface ChangePatronProps {
  (newValue: string | boolean, key: string): void;
}
