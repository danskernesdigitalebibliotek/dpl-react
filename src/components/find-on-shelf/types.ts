import { HoldingsLogisticsV1 } from "../../core/fbs/model";
import { Manifestation } from "../../core/utils/types/entities";

// This type is preferrably supposed to be used for matching manifestation +
// holding pairs so the holding corresponds to the manifestation.
export type ManifestationHoldings = {
  manifestation: Manifestation;
  holding: HoldingsLogisticsV1;
}[];

export type SelectedPeriodicalEdition = {
  selectedYear: string;
  selectedEdition: string;
};
