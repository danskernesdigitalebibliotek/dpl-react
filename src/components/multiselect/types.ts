export type MultiselectOption = {
  item: string;
  value: string;
};

export type MultiselectExternalUpdateFunction = (filtersUpdate: {
  key: string;
  value: MultiselectOption[];
}) => void;

export type MultiselectExternalUpdate = {
  key: string;
  externalUpdateFunction: MultiselectExternalUpdateFunction;
};
