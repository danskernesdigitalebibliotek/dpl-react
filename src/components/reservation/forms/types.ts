export type FormSelectValue = string | number;
