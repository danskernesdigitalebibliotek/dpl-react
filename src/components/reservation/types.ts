import { Option } from "../Dropdown/Dropdown";

export type InterestPeriods = {
  interestPeriods: Option[];
  defaultInterestPeriod: Option;
};
