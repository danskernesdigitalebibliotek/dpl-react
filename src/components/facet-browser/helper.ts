import { mapValues } from "lodash";
import {
  FacetFieldEnum,
  FacetResult,
  FacetValue,
  useSearchFacetQuery
} from "../../core/dbc-gateway/generated/graphql";
import useGetCleanBranches from "../../core/utils/branches";
import { Filter, FilterItemTerm } from "../../core/filter.slice";
import invalidSwitchCase from "../../core/utils/helpers/invalid-switch-case";
import { Facets } from "../../core/utils/types/entities";

export const allFacetFields = [
  FacetFieldEnum.Materialtypesgeneral,
  FacetFieldEnum.Creators,
  FacetFieldEnum.Subjects,
  FacetFieldEnum.Fictionnonfiction,
  FacetFieldEnum.Childrenoradults,
  FacetFieldEnum.Accesstypes,
  FacetFieldEnum.Mainlanguages,
  FacetFieldEnum.Genreandform,
  FacetFieldEnum.Materialtypesspecific,
  FacetFieldEnum.Fictionalcharacters
];

export const getPlaceHolderFacets = (facets: string[]) =>
  facets.map((facet) => ({
    name: facet,
    type: facet as FacetFieldEnum,
    values: [
      {
        key: "",
        term: ""
      }
    ]
  }));

export const formatFacetTerms = (filters: {
  [key: string]: { [key: string]: FilterItemTerm };
}) => {
  return Object.keys(filters).reduce(
    (acc, key) => ({
      ...acc,
      [key]: Object.keys(filters[key])
    }),
    {}
  );
};

export const createFilters = (
  facets: {
    [key: string]: { [key: string]: FilterItemTerm };
  },
  branchIdList: string[]
) => {
  return {
    ...formatFacetTerms(facets),
    ...(branchIdList ? { branchId: branchIdList } : {})
  };
};

export function useGetFacets(query: string, filters: Filter) {
  const cleanBranches = useGetCleanBranches();

  const { data, isLoading } = useSearchFacetQuery(
    {
      q: { all: query },
      facets: allFacetFields,
      facetLimit: 10,
      filters: createFilters(filters, cleanBranches)
    },
    {
      keepPreviousData: true,
      placeholderData: {
        search: {
          facets: getPlaceHolderFacets(allFacetFields)
        }
      }
    }
  );

  return { facets: data?.search.facets || null, isLoading };
}

export const FacetBrowserModalId = "facet-browser-modal";

export function getAllFilterPathsAsString(filterObject: {
  [key: string]: { [key: string]: FilterItemTerm };
}) {
  const mappedFilterValues = mapValues(filterObject, (filter) => {
    return Object.keys(filter);
  });
  const filterNames = Object.keys(mappedFilterValues);
  let allFilterPathsAsString = "";
  filterNames.forEach((filterName) => {
    mappedFilterValues[filterName].forEach((filterValue) => {
      if (allFilterPathsAsString !== "") {
        allFilterPathsAsString = allFilterPathsAsString.concat(";");
      }
      allFilterPathsAsString = allFilterPathsAsString.concat(
        `facet.${filterName}:${filterValue}`
      );
    });
  });
  return allFilterPathsAsString;
}

export const getFacetFieldTranslation = (name: FacetFieldEnum) => {
  switch (name.toLowerCase()) {
    case FacetFieldEnum.Accesstypes.toLowerCase():
      return "facetAccessTypesText";
    case FacetFieldEnum.Canalwaysbeloaned.toLowerCase():
      return "facetCanAlwaysBeLoanedText";
    case FacetFieldEnum.Childrenoradults.toLowerCase():
      return "facetChildrenOrAdultsText";
    case FacetFieldEnum.Creators.toLowerCase():
      return "facetCreatorsText";
    case FacetFieldEnum.Dk5.toLowerCase():
      return "facetDk5Text";
    case FacetFieldEnum.Fictionnonfiction.toLowerCase():
      return "facetFictionNonfictionText";
    case FacetFieldEnum.Fictionalcharacters.toLowerCase():
      return "facetFictionalCharactersText";
    case FacetFieldEnum.Genreandform.toLowerCase():
      return "facetGenreAndFormText";
    case FacetFieldEnum.Mainlanguages.toLowerCase():
      return "facetMainLanguagesText";
    case FacetFieldEnum.Materialtypesspecific.toLowerCase():
      return "facetMaterialTypesGeneralText";
    case FacetFieldEnum.Materialtypesgeneral.toLowerCase():
      return "facetMaterialTypesSpecificText";
    case FacetFieldEnum.Subjects.toLowerCase():
      return "facetSubjectsText";
    case FacetFieldEnum.Worktypes.toLowerCase():
      return "facetWorkTypesText";
    case FacetFieldEnum.Year.toLowerCase():
      return "facetYearText";
    case FacetFieldEnum.Gameplatform.toLowerCase():
      return "facetGamePlatformText";
    default:
      return invalidSwitchCase<string>(name as never);
  }
};

type FacetMap = { [key: string]: FacetValue };
// createFacetsMap generates a map for quick lookup and enhanced search
// capabilities by combining facet name and term into a single string key.
// Structure: { [facetName:termKey]: FacetValue }
// Example Key: 'fictionalCharacters:Batman'
export const createFacetsMap = (facets: FacetResult[]): FacetMap => {
  return facets.reduce((acc: FacetMap, facet: FacetResult) => {
    const newAcc = { ...acc };

    facet.values.forEach((value) => {
      const combinedKey = `${facet.name}:${value.key}`;
      newAcc[combinedKey] = value;
    });

    return newAcc;
  }, {});
};

export const findTermInFacetMap = (
  facetName: string,
  termName: string,
  facetMap: FacetMap
): FacetValue => {
  const key = `${facetName}:${termName}`;
  return facetMap[key];
};

export default {};

/* ********************************* Vitest Section  ********************************* */
if (import.meta.vitest) {
  const { describe, expect, it } = import.meta.vitest;

  const branchIdList = ["710100", "710200"];

  const filters = {
    mainLanguages: {
      Dansk: {
        key: "dan",
        term: "Dansk",
        score: 295
      },
      Engelsk: {
        key: "eng",
        term: "Engelsk",
        score: 9
      }
    },
    accessTypes: {
      Fysisk: {
        key: "PHYSICAL",
        term: "Fysisk",
        score: 356
      }
    }
  };

  const facetsTestData: Facets = [
    {
      name: "materialTypesGeneral",
      type: FacetFieldEnum.Materialtypesgeneral,
      values: [
        {
          key: "computerspil",
          term: "computerspil",
          score: 26
        }
      ]
    },
    {
      name: "mainLanguages",
      type: FacetFieldEnum.Mainlanguages,
      values: [
        {
          key: "dan",
          term: "Dansk",
          score: 427
        },
        {
          key: "eng",
          term: "Engelsk",
          score: 82
        }
      ]
    },
    {
      name: "materialTypesGeneral",
      type: FacetFieldEnum.Materialtypesgeneral,
      values: [
        {
          key: "artikler",
          term: "artikler",
          score: 346
        },
        {
          key: "bøger",
          term: "bøger",
          score: 74
        },
        {
          key: "e-bøger",
          term: "e-bøger",
          score: 38
        },
        {
          key: "computerspil",
          term: "computerspil",
          score: 26
        },
        {
          key: "film",
          term: "film",
          score: 9
        }
      ]
    }
  ];

  describe("getPlaceHolderFacets", () => {
    it("should get placeholder facets", () => {
      expect(getPlaceHolderFacets(allFacetFields)).toMatchSnapshot();
    });
  });

  describe("formatFacetTerms", () => {
    it("should format facet terms", () => {
      expect(formatFacetTerms(filters)).toMatchSnapshot();
    });
  });

  describe("createFilters", () => {
    it("should create filters", () => {
      expect(createFilters(filters, branchIdList)).toMatchSnapshot();
    });
  });

  it("createFacetsMap", () => {
    expect(createFacetsMap(facetsTestData)).toMatchSnapshot();
  });
}
