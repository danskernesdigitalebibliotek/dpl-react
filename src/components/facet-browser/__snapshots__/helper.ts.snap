// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createFacetsMap 1`] = `
{
  "mainLanguages:dan": {
    "key": "dan",
    "score": 427,
    "term": "Dansk",
  },
  "mainLanguages:eng": {
    "key": "eng",
    "score": 82,
    "term": "Engelsk",
  },
  "materialTypesGeneral:artikler": {
    "key": "artikler",
    "score": 346,
    "term": "artikler",
  },
  "materialTypesGeneral:bøger": {
    "key": "bøger",
    "score": 74,
    "term": "bøger",
  },
  "materialTypesGeneral:computerspil": {
    "key": "computerspil",
    "score": 26,
    "term": "computerspil",
  },
  "materialTypesGeneral:e-bøger": {
    "key": "e-bøger",
    "score": 38,
    "term": "e-bøger",
  },
  "materialTypesGeneral:film": {
    "key": "film",
    "score": 9,
    "term": "film",
  },
}
`;

exports[`createFilters > should create filters 1`] = `
{
  "accessTypes": [
    "Fysisk",
  ],
  "branchId": [
    "710100",
    "710200",
  ],
  "mainLanguages": [
    "Dansk",
    "Engelsk",
  ],
}
`;

exports[`formatFacetTerms > should format facet terms 1`] = `
{
  "accessTypes": [
    "Fysisk",
  ],
  "mainLanguages": [
    "Dansk",
    "Engelsk",
  ],
}
`;

exports[`getPlaceHolderFacets > should get placeholder facets 1`] = `
[
  {
    "name": "MATERIALTYPESGENERAL",
    "type": "MATERIALTYPESGENERAL",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "CREATORS",
    "type": "CREATORS",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "SUBJECTS",
    "type": "SUBJECTS",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "FICTIONNONFICTION",
    "type": "FICTIONNONFICTION",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "CHILDRENORADULTS",
    "type": "CHILDRENORADULTS",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "ACCESSTYPES",
    "type": "ACCESSTYPES",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "MAINLANGUAGES",
    "type": "MAINLANGUAGES",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "GENREANDFORM",
    "type": "GENREANDFORM",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "MATERIALTYPESSPECIFIC",
    "type": "MATERIALTYPESSPECIFIC",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
  {
    "name": "FICTIONALCHARACTERS",
    "type": "FICTIONALCHARACTERS",
    "values": [
      {
        "key": "",
        "term": "",
      },
    ],
  },
]
`;
